# Lecture 5 — Tutorial 3.14 Applied Activities

**Goal.** Complete eight cumulative exploration activities before opening the separated model walkthrough.

This notebook matches the numbered Lecture 5 website tutorials. The dataset is synthetic, contains deliberate quality problems and must not be interpreted as evidence about real cities or services.

**Local setup.** Google Colab includes pandas. On a local machine, run `python -m pip install -r requirements.txt` from the repository, or install pandas with `python -m pip install "pandas>=2.0"`.


## Setup

Run this cell first. It loads the versioned raw CSV from GitHub and checks the expected 121 × 14 structure. Keep the raw data unchanged during Lecture 5.


In [1]:
import pandas as pd  # Loads the library feature used by this step.
data_url = "https://raw.githubusercontent.com/asmrabbi/E26_TAN7_Scripting_CPH/main/data/E26_TAN7_service_experience_raw.csv"  # Stores the calculated or supplied value for later use.
df = pd.read_csv(data_url)  # Stores the calculated or supplied value for later use.
print("Dataset shape:", df.shape)  # Displays the result so it can be inspected.
assert df.shape == (121, 14), "The wrong dataset version was loaded."  # Stores the calculated or supplied value for later use.
display(df.head())  # Displays the result so it can be inspected.


Dataset shape: (121, 14)
     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.
3  SR2026-0004  ...  Helpful guidance, but some users reported dela...
4  SR2026-0005  ...  The response was fast and the information was ...

[5 rows x 14 columns]


# Student applied activities

Complete the activities in order and write your own report. The model walkthrough begins only after the STOP divider.


## Applied Activity 3.14.1 — Identify and describe the dataset

### Your task

Write a short dataset identity section that states:

1. dataset name;
2. source/location;
3. whether the data are real or synthetic;
4. intended unit of observation;
5. number of rows and columns;
6. all column names;
7. which columns appear to be identifiers, categorical, numeric, date-like and free text.

### Required code

At minimum use:

```python
df.shape
df.columns
df.dtypes
df.info()
```

### Required written output

Write 4–6 sentences. Do not simply paste the code output.



---


In [2]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.2 — Inspect the structure and sample records

### Your task

Inspect:

- first five rows;
- final five rows;
- five sampled rows;
- shape;
- non-missing counts;
- dtypes.

Use a reproducible sample:

```python
df.sample(5, random_state=42)
```

Then answer:

1. Which columns look straightforwardly numeric?
2. Which columns need closer inspection before numeric/date analysis?
3. Why is inspecting only `head()` insufficient?

### Suggested code

```python
print("Shape:", df.shape)

display(df.head())
display(df.tail())
display(df.sample(5, random_state=42))

print(df.dtypes)
df.info()
```

---


In [3]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.3 — Calculate preliminary numerical summaries

### Your task

For `cases_received`, calculate:

- count;
- sum;
- mean;
- median;
- standard deviation;
- minimum;
- maximum;
- `describe()`.

For `cases_resolved`, calculate:

- sum;
- mean;
- median.

For `satisfaction_score`, calculate:

- non-missing count;
- mean;
- median;
- minimum;
- maximum.

### Starter code

```python
print("Cases received")
print("Count:", df["cases_received"].count())
print("Sum:", df["cases_received"].sum())
print("Mean:", df["cases_received"].mean())
print("Median:", df["cases_received"].median())
print("Std:", df["cases_received"].std())
print("Min:", df["cases_received"].min())
print("Max:", df["cases_received"].max())

print("\nFull summary")
print(df["cases_received"].describe())
```

### Interpretation questions

1. Which result immediately suggests a possible validity problem?
2. Why should the raw mean not yet be described as a final result?
3. How many satisfaction scores contribute to the raw mean?
4. Does the raw satisfaction maximum fit the documented 1–5 scale?



---


In [4]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.4 — Create and inspect derived measures

### Your task

Create:

```text
unresolved_cases = cases_received - cases_resolved
```

and:

```text
resolution_rate = cases_resolved / cases_received × 100
```

Then inspect:

- negative unresolved values;
- rates above 100%;
- missing/undefined rates caused by a zero denominator.

### Starter code

```python
df["unresolved_cases"] = (
    df["cases_received"] - df["cases_resolved"]
)

df["resolution_rate"] = (
    df["cases_resolved"] /
    df["cases_received"] * 100
)

columns = [
    "record_id",
    "cases_received",
    "cases_resolved",
    "unresolved_cases",
    "resolution_rate"
]

display(df[columns])
```

### Additional calculations

Calculate both:

```python
mean_row_rate = df["resolution_rate"].mean()
```

and:

```python
overall_rate = (
    df["cases_resolved"].sum() /
    df["cases_received"].sum() * 100
)
```

### Questions

1. Why do these two results differ?
2. Which denominator does each method use conceptually?
3. Why should neither raw result yet be treated as final?



---


In [5]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.5 — Explore categories and group differences

### Your task

For `service_type`:

1. print unique values;
2. calculate counts;
3. calculate percentages;
4. calculate mean and median `cases_received` by service type;
5. calculate count, mean, median, minimum and maximum `satisfaction_score` by service type.

For `city`:

1. print unique raw labels;
2. calculate raw frequencies;
3. calculate grouped mean `cases_received` by raw city label;
4. explain why the result should not yet be interpreted as a clean city comparison.

Create a crosstab of:

```text
city × service_type
```

### Suggested code

```python
print(df["service_type"].value_counts())
print(df["service_type"].value_counts(normalize=True).mul(100).round(2))

service_cases = (
    df.groupby("service_type")["cases_received"]
      .agg(["count", "sum", "mean", "median", "min", "max"])
)

display(service_cases.round(2))

service_satisfaction = (
    df.groupby("service_type")["satisfaction_score"]
      .agg(["count", "mean", "median", "min", "max"])
)

display(service_satisfaction.round(2))

print(df["city"].unique())
print(df["city"].value_counts())

city_service = pd.crosstab(
    df["city"],
    df["service_type"]
)

display(city_service)
```

### Required interpretation

Write at least two preliminary observations and then explicitly state which raw-data problems could distort them.



---


In [6]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.6 — Complete a structured data-quality audit

### Your task

Create checks for:

1. missing values by column;
2. missing percentages;
3. exact duplicate rows;
4. repeated `record_id` values;
5. raw city-category inconsistencies;
6. negative `cases_received`;
7. satisfaction outside 1–5;
8. `cases_resolved > cases_received`;
9. non-numeric `resolution_days`;
10. invalid `report_month` values.

### Suggested audit code

```python
print("Missing values")
print(df.isna().sum())

print("\nMissing percentages")
print((df.isna().mean() * 100).round(2))

print("\nExact duplicates")
print(df.duplicated().sum())
display(df[df.duplicated(keep=False)])

print("\nRepeated IDs")
id_duplicate_mask = df["record_id"].duplicated(keep=False)
display(df[id_duplicate_mask])

print("\nRaw city labels")
print(df["city"].value_counts(dropna=False))

print("\nNegative received cases")
display(df[df["cases_received"] < 0])

print("\nOut-of-range satisfaction")
sat_problem = (
    (df["satisfaction_score"] < 1) |
    (df["satisfaction_score"] > 5)
)
display(df[sat_problem])

print("\nResolved greater than received")
display(df[df["cases_resolved"] > df["cases_received"]])

resolution_preview = pd.to_numeric(
    df["resolution_days"],
    errors="coerce"
)
resolution_problem = (
    df["resolution_days"].notna() &
    resolution_preview.isna()
)
print("\nNon-numeric resolution values")
display(df.loc[resolution_problem, ["record_id", "resolution_days"]])

month_preview = pd.to_datetime(
    df["report_month"],
    errors="coerce"
)
print("\nInvalid report months")
display(df.loc[month_preview.isna(), ["record_id", "report_month"]])
```

### Required output: quality log

Create a Markdown table with at least these columns:

| Issue | How detected | Affected variable/record | Why it matters | What should happen next? |
|---|---|---|---|---|

Do not put “delete” or “replace” as an automatic next step unless you first explain the evidence required to justify that action.

---


In [7]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.7 — Explore the feedback qualitatively and computationally

### Your task

1. Display all non-missing feedback.
2. Read the comments manually.
3. Identify at least three recurring ideas.
4. Create at least two transparent Boolean keyword indicators.
5. Count each indicator.
6. Display the matching comments.
7. Compare one indicator with `service_type` or `satisfaction_score`.
8. Describe at least one false-positive/false-negative risk.

### Example indicator setup

```python
df["mentions_wait"] = df["feedback"].str.contains(
    "wait|long",
    case=False,
    na=False,
    regex=True
)

df["mentions_helpful"] = df["feedback"].str.contains(
    "helpful",
    case=False,
    na=False
)
```

### Count the rules

```python
print("Waiting matches:", df["mentions_wait"].sum())
print("Helpful matches:", df["mentions_helpful"].sum())
```

Expected raw counts:

```text
Waiting matches: 11
Helpful matches: 36
```

### Compare waiting indicator and service type

```python
wait_by_service = pd.crosstab(
    df["service_type"],
    df["mentions_wait"]
)

display(wait_by_service)
```

### Compare waiting indicator and satisfaction

```python
wait_score = (
    df.groupby("mentions_wait")["satisfaction_score"]
      .agg(["count", "mean", "median"])
)

display(wait_score.round(2))
```

### Required written interpretation

Explain why this is a **descriptive association based on a simple coding rule**, not evidence that waiting causes satisfaction outcomes.

Also check how many waiting-related comments have a non-missing satisfaction score, so the number of comments matching the text rule and the number contributing to the score average are different.

---


In [8]:
pass  # Runs this part of the documented analysis.


## Applied Activity 3.14.8 — Write the Initial Data Exploration Report

Create a short report with these headings.

### A. Dataset and provenance

Include:

- dataset name;
- source;
- synthetic status;
- unit of observation.

### B. Structure

Include:

- number of rows/columns;
- key variables;
- conceptual types.

### C. Preliminary quantitative exploration

Include at least:

- one total;
- one mean;
- one median;
- one category percentage;
- one grouped comparison;
- one derived measure.

Every result must state what it counts/measures.

### D. Data-quality concerns

Include at least five identified issues and explain why they matter.

### E. Text/qualitative exploration

Include:

- manually observed ideas;
- at least one computational coding rule;
- match count;
- limitation of the rule.

### F. Three preliminary findings

Each finding should contain:

```text
WHAT WAS OBSERVED
+
HOW IT WAS CALCULATED/FOUND
+
WHAT LIMITS THE INTERPRETATION
```

### G. Next steps for Lecture 6

Identify which issues require:

- category standardisation;
- type conversion;
- date handling;
- missing-data decisions;
- duplicate verification;
- range validation;
- text preparation.

---

<details>
<summary>Open the Tutorial 3.14 model walkthrough</summary>


In [9]:
pass  # Runs this part of the documented analysis.


# STOP — Model walkthrough begins below

Compare the following reproducible checkpoints only after completing your own analysis and written interpretation.


## Model 3.14.1 — Identify and describe the dataset

The supplied file is synthetic teaching data. One row is intended to represent one service-category-city-month record, but that intended grain should be checked because one exact duplicate exists. The raw table has 121 rows and 14 columns. Technical dtypes and conceptual roles differ for `report_month` and `resolution_days`.


In [10]:
print("Shape:", df.shape)  # Displays the result so it can be inspected.
print("Columns:", df.columns.tolist())  # Displays the result so it can be inspected.
print(df.dtypes)  # Displays the result so it can be inspected.
df.info()  # Runs this part of the documented analysis.


Shape: (121, 14)
Columns: ['record_id', 'report_month', 'city', 'service_type', 'cases_received', 'cases_resolved', 'resolution_days', 'satisfaction_score', 'digital_cases', 'in_person_cases', 'repeat_contacts', 'complaints_received', 'staff_hours', 'feedback']
record_id               object
report_month            object
city                    object
service_type            object
cases_received           int64
cases_resolved           int64
resolution_days         object
satisfaction_score     float64
digital_cases            int64
in_person_cases          int64
repeat_contacts          int64
complaints_received      int64
staff_hours            float64
feedback                object
dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121 entries, 0 to 120
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   record_id            121 non-null    object 
 1   report_month         121 non-n

## Model 3.14.2 — Inspect the structure and sample records

The samples show straightforward numeric counts and measures, categorical labels, a date-like string, an identifier and free text. `resolution_days` and `report_month` require validation. `head()` alone can miss rare values, later source systems and deliberate anomalies.


In [11]:
print("Shape:", df.shape)  # Displays the result so it can be inspected.
display(df.head())  # Displays the result so it can be inspected.
display(df.tail())  # Displays the result so it can be inspected.
display(df.sample(5, random_state=42))  # Displays the result so it can be inspected.
print(df.dtypes)  # Displays the result so it can be inspected.
df.info()  # Runs this part of the documented analysis.


Shape: (121, 14)
     record_id  ...                                           feedback
0  SR2026-0001  ...  The guidance was clear, but the supporting-doc...
1  SR2026-0002  ...  Several users asked for clearer information ab...
2  SR2026-0003  ...  Useful advice and a friendly response from staff.
3  SR2026-0004  ...  Helpful guidance, but some users reported dela...
4  SR2026-0005  ...  The response was fast and the information was ...

[5 rows x 14 columns]
       record_id  ...                                           feedback
116  SR2026-0117  ...  The digital form worked well, although some in...
117  SR2026-0118  ...                                                NaN
118  SR2026-0119  ...  Collection information was easy to understand ...
119  SR2026-0120  ...  The response was fast and the information was ...
120  SR2026-0027  ...  The digital form worked well, although some in...

[5 rows x 14 columns]
      record_id  ...                                           feedback
4

## Model 3.14.3 — Calculate preliminary numerical summaries

The negative received count and the 0.0/6.2 satisfaction values conflict with the documented rules. Exactly 117 satisfaction scores contribute to the raw mean. All summaries remain preliminary until the raw problems are investigated and a cleaning log is approved.


In [12]:
resolved = df["cases_resolved"]  # Stores the calculated or supplied value for later use.
print("Count:", resolved.count())  # Displays the result so it can be inspected.
print("Sum:", resolved.sum())  # Displays the result so it can be inspected.
print("Mean:", resolved.mean())  # Displays the result so it can be inspected.
print("Median:", resolved.median())  # Displays the result so it can be inspected.
print("Minimum:", resolved.min())  # Displays the result so it can be inspected.
print("Maximum:", resolved.max())  # Displays the result so it can be inspected.
print("Standard deviation:", resolved.std())  # Displays the result so it can be inspected.
print(resolved.describe())  # Displays the result so it can be inspected.
print(df["cases_received"].describe())  # Displays the result so it can be inspected.
print("Satisfaction count:", df["satisfaction_score"].count())  # Displays the result so it can be inspected.
print("Satisfaction mean:", df["satisfaction_score"].mean())  # Displays the result so it can be inspected.
print("Satisfaction median:", df["satisfaction_score"].median())  # Displays the result so it can be inspected.
print("Satisfaction minimum:", df["satisfaction_score"].min())  # Displays the result so it can be inspected.
print("Satisfaction maximum:", df["satisfaction_score"].max())  # Displays the result so it can be inspected.


Count: 121
Sum: 11958
Mean: 98.82644628099173
Median: 95.0
Minimum: 0
Maximum: 176
Standard deviation: 26.195953149914338
count    121.000000
mean      98.826446
std       26.195953
min        0.000000
25%       82.000000
50%       95.000000
75%      113.000000
max      176.000000
Name: cases_resolved, dtype: float64
count    121.000000
mean     104.570248
std       27.256750
min       -4.000000
25%       86.000000
50%      102.000000
75%      122.000000
max      185.000000
Name: cases_received, dtype: float64
Satisfaction count: 117
Satisfaction mean: 4.112820512820512
Satisfaction median: 4.2
Satisfaction minimum: 0.0
Satisfaction maximum: 6.2


## Model 3.14.4 — Create and inspect derived measures

The mean row rate gives every record equal weight, while the aggregate rate weights records through their case volumes. Three rows exceed 100% and one negative denominator makes a row result substantively invalid. Preserve these flags rather than hiding them.


In [13]:
df["unresolved_cases"] = df["cases_received"] - df["cases_resolved"]  # Stores the calculated or supplied value for later use.
print(df["unresolved_cases"].nsmallest(5))  # Displays the result so it can be inspected.
display(df[df["unresolved_cases"] < 0][["record_id", "cases_received", "cases_resolved", "unresolved_cases"]])  # Displays the result so it can be inspected.
df["resolution_rate"] = df["cases_resolved"] / df["cases_received"] * 100  # Stores the calculated or supplied value for later use.
display(df[df["resolution_rate"] > 100][["record_id", "resolution_rate"]])  # Displays the result so it can be inspected.
print("Mean row rate:", round(df["resolution_rate"].mean(), 2))  # Displays the result so it can be inspected.
print("Aggregate rate:", round(df["cases_resolved"].sum() / df["cases_received"].sum() * 100, 2))  # Displays the result so it can be inspected.


11   -7
53   -4
89   -3
79    1
91    1
Name: unresolved_cases, dtype: int64
      record_id  cases_received  cases_resolved  unresolved_cases
11  SR2026-0012              80              87                -7
53  SR2026-0054              -4               0                -4
89  SR2026-0090             159             162                -3
      record_id  resolution_rate
11  SR2026-0012       108.750000
89  SR2026-0090       101.886792
Mean row rate: 93.66
Aggregate rate: 94.51


## Model 3.14.5 — Explore categories and group differences

Exact raw labels split intended categories, so grouped output contains nine service strings instead of five and several city variants. Group summaries are useful diagnostics but not clean comparisons. The unit remains records, not citizens.


In [14]:
print(df["service_type"].unique())  # Displays the result so it can be inspected.
print("Distinct raw labels:", df["service_type"].nunique())  # Displays the result so it can be inspected.
print(df["service_type"].value_counts())  # Displays the result so it can be inspected.
print(df["service_type"].value_counts(normalize=True).mul(100).round(2))  # Displays the result so it can be inspected.
print("Modes:", df["service_type"].mode().tolist())  # Displays the result so it can be inspected.
service_summary = df.groupby("service_type")["cases_received"].agg(["count", "mean", "median", "min", "max"])  # Stores the calculated or supplied value for later use.
display(service_summary.round(2))  # Displays the result so it can be inspected.
service_satisfaction = df.groupby("service_type")["satisfaction_score"].agg(["count", "mean", "median", "min", "max"])  # Stores the calculated or supplied value for later use.
display(service_satisfaction.round(2))  # Displays the result so it can be inspected.
print(df["city"].value_counts())  # Displays the result so it can be inspected.
display(pd.crosstab(df["city"], df["service_type"]))  # Displays the result so it can be inspected.


['Housing' 'Transport' 'Employment' 'Waste' 'Citizen Services'
 'Transport ' 'Citizen services' 'WASTE' 'Employment Services']
Distinct raw labels: 9
service_type
Transport              24
Employment             24
Waste                  24
Citizen Services       24
Housing                21
Transport               1
Citizen services        1
WASTE                   1
Employment Services     1
Name: count, dtype: int64
service_type
Transport              19.83
Employment             19.83
Waste                  19.83
Citizen Services       19.83
Housing                17.36
Transport               0.83
Citizen services        0.83
WASTE                   0.83
Employment Services     0.83
Name: proportion, dtype: float64
Modes: ['Citizen Services', 'Employment', 'Transport', 'Waste']
                     count    mean  median  min  max
service_type                                        
Citizen Services        24  129.92   122.0   94  185
Citizen services         1   99.00    99.0   99

## Model 3.14.6 — Complete a structured data-quality audit

A complete quality log should record the rule, affected IDs, observed value, why it matters and the evidence needed before repair. Detection is not permission to delete, clip, merge or impute.


In [15]:
print("Missing counts")  # Displays the result so it can be inspected.
print(df.isna().sum())  # Displays the result so it can be inspected.
print("Missing percentages")  # Displays the result so it can be inspected.
print(df.isna().mean().mul(100).round(2))  # Displays the result so it can be inspected.
print("Exact duplicates:", df.duplicated().sum())  # Displays the result so it can be inspected.
display(df[df.duplicated(keep=False)])  # Displays the result so it can be inspected.
display(df[df["record_id"].duplicated(keep=False)])  # Displays the result so it can be inspected.
print("Raw cities:", df["city"].unique().tolist())  # Displays the result so it can be inspected.
display(df[df["cases_received"] < 0])  # Displays the result so it can be inspected.
sat_problem = (df["satisfaction_score"] < 1) | (df["satisfaction_score"] > 5)  # Stores the calculated or supplied value for later use.
display(df[sat_problem])  # Displays the result so it can be inspected.
display(df[df["cases_resolved"] > df["cases_received"]])  # Displays the result so it can be inspected.
resolution_preview = pd.to_numeric(df["resolution_days"], errors="coerce")  # Stores the calculated or supplied value for later use.
display(df.loc[df["resolution_days"].notna() & resolution_preview.isna(), ["record_id", "resolution_days"]])  # Displays the result so it can be inspected.
month_preview = pd.to_datetime(df["report_month"], errors="coerce")  # Stores the calculated or supplied value for later use.
display(df.loc[month_preview.isna(), ["record_id", "report_month"]])  # Displays the result so it can be inspected.
print("Additional cross-field checks")  # Displays the result so it can be inspected.
print("Digital plus in-person mismatch:", int(((df["digital_cases"] + df["in_person_cases"]) != df["cases_received"]).sum()))  # Displays the result so it can be inspected.
print("Negative complaints:", int((df["complaints_received"] < 0).sum()))  # Displays the result so it can be inspected.
print("Non-positive staff hours:", int((df["staff_hours"] <= 0).sum()))  # Displays the result so it can be inspected.


Missing counts
record_id              0
report_month           0
city                   0
service_type           0
cases_received         0
cases_resolved         0
resolution_days        0
satisfaction_score     4
digital_cases          0
in_person_cases        0
repeat_contacts        0
complaints_received    0
staff_hours            0
feedback               3
unresolved_cases       0
resolution_rate        0
dtype: int64
Missing percentages
record_id              0.00
report_month           0.00
city                   0.00
service_type           0.00
cases_received         0.00
cases_resolved         0.00
resolution_days        0.00
satisfaction_score     3.31
digital_cases          0.00
in_person_cases        0.00
repeat_contacts        0.00
complaints_received    0.00
staff_hours            0.00
feedback               2.48
unresolved_cases       0.00
resolution_rate        0.00
dtype: float64
Exact duplicates: 1
       record_id report_month  ... unresolved_cases resolution_rate
2

## Model 3.14.7 — Explore the feedback qualitatively and computationally

Keyword indicators retrieve comments matching explicit text rules; they are not complete qualitative themes. Context checking is necessary because synonyms, negation and mixed comments can create false negatives or false positives. Associations with scores do not prove causation.


In [16]:
df["mentions_wait"] = df["feedback"].str.contains("wait|long", case=False, na=False, regex=True)  # Stores the calculated or supplied value for later use.
df["mentions_helpful"] = df["feedback"].str.contains("helpful", case=False, na=False)  # Stores the calculated or supplied value for later use.
print("Waiting matches:", int(df["mentions_wait"].sum()))  # Displays the result so it can be inspected.
print("Helpful matches:", int(df["mentions_helpful"].sum()))  # Displays the result so it can be inspected.
display(pd.crosstab(df["service_type"], df["mentions_wait"]))  # Displays the result so it can be inspected.
wait_score = df.groupby("mentions_wait")["satisfaction_score"].agg(["count", "mean", "median"])  # Stores the calculated or supplied value for later use.
display(wait_score.round(2))  # Displays the result so it can be inspected.
print("Waiting comments with scores:", df.loc[df["mentions_wait"], "satisfaction_score"].notna().sum())  # Displays the result so it can be inspected.


Waiting matches: 11
Helpful matches: 36
mentions_wait        False  True 
service_type                     
Citizen Services        24      0
Citizen services         1      0
Employment              19      5
Employment Services      1      0
Housing                 15      6
Transport               24      0
Transport                1      0
WASTE                    1      0
Waste                   24      0
               count  mean  median
mentions_wait                     
False            107  4.18    4.20
True              10  3.39    3.65
Waiting comments with scores: 10


## Model 3.14.8 — Write the Initial Data Exploration Report

Use the printed checkpoints to write the report headings on the website: provenance, structure, quantitative exploration, quality concerns, text exploration, limitations and next steps. Every number must state its unit and denominator, and the conclusion must keep real-world generalisability at none.


In [17]:
report_checkpoints = {"rows": len(df), "columns": df.shape[1], "total_cases_received": int(df["cases_received"].sum()), "mean_cases_received": round(df["cases_received"].mean(), 2), "median_cases_received": float(df["cases_received"].median()), "missing_cells": int(df.isna().sum().sum()), "exact_duplicate_rows": int(df.duplicated().sum()), "invalid_satisfaction": int(((df["satisfaction_score"] < 1) | (df["satisfaction_score"] > 5)).sum())}  # Stores the calculated or supplied value for later use.
for label, value in report_checkpoints.items():  # Starts a loop over the relevant values.
    print(label, "=", value)  # Displays the result so it can be inspected.


rows = 121
columns = 18
total_cases_received = 12653
mean_cases_received = 104.57
median_cases_received = 102.0
missing_cells = 7
exact_duplicate_rows = 1
invalid_satisfaction = 2


# Checks and next steps

All eight website activities and eight model sections are present. Save your report, keep the raw file unchanged and carry the documented quality log into the Lecture 6 cleaning workflow.
